compiler/passes/convert-uast.cpp

```diff
@@ -314,6 +315,106 @@ struct Converter {
+    if (node->numActuals() != 1 && node->numActuals() != 2) return nullptr;
+
+    if (!node->actual(0)->isStringLiteral()) return nullptr;
+    auto attrName = node->actual(0)->toStringLiteral()->value().astr(context);
+
+    if (node->numActuals() == 1) {
+      return LLVMMetadata::construct(attrName);
+    } else {
+      auto attrVal = node->actual(1);
+
+      if (auto str = attrVal->toStringLiteral())
+        return LLVMMetadata::constructString(attrName, str->value().astr(context));
+      else if (auto int_ = attrVal->toIntLiteral())
+        return LLVMMetadata::constructInt(attrName, int_->value());
+      else if (auto bool_ = attrVal->toBoolLiteral())
+        return LLVMMetadata::constructBool(attrName, bool_->value());
+      else if (auto tup = attrVal->toTuple()) {
+        auto v = tupleToLLVMMetadata(tup);
+        if (v == nullptr) return nullptr;
+        return LLVMMetadata::constructMetadata(attrName, v);
+      }
+      else return nullptr;
+    }
+  }
+
+  LLVMMetadataPtr nodeToLLVMMetadata(const uast::AstNode* node) {
+    if (node->isTuple()) {
+      return tupleToLLVMMetadata(node->toTuple());
+    } else if (node->isStringLiteral()) {
+      auto attrName = node->toStringLiteral()->value().astr(context);
+      return LLVMMetadata::construct(attrName);
+    } else {
+    return nullptr;
+    }
+  }
+
+  LLVMMetadataList buildLLVMMetadataList(const uast::AttributeGroup* node) {
+    LLVMMetadataList llvmAttrs;
+    if(!node) return llvmAttrs;
+
+    if (auto llvmAttrNode = node->getAttributeNamed(UniqueString::get(context, "llvm.metadata"))) {
+
+      // no matter what, warn for this attribute that its unstable
+      auto attrLoc = chpl::parsing::locateId(context, llvmAttrNode->id());
+      bool warnUnstable = chpl::parsing::shouldWarnUnstableForPath(context, attrLoc.path());
+      if (warnUnstable) {
+        // can't use CHPL_REPORT or reportUnstableWarningForId
+        // because AttributeGroup is not a NamedDecl
+        std::string msg = "'llvm.metadata' is an unstable attribute";
+        auto err = GeneralError::get(ErrorBase::WARNING, attrLoc, msg);
+        context->report(std::move(err));
+      }
+
+      for (auto act: llvmAttrNode->actuals()) {
+        auto attr = nodeToLLVMMetadata(act);
+        if (attr != nullptr) {
+          llvmAttrs.push_back(attr);
+        } else {
+          std::string msg = "Invalid value for 'llvm.metadata'";
+          auto err = GeneralError::get(ErrorBase::ERROR, attrLoc, msg);
+          context->report(std::move(err));
+        }
+      }
+    }
+    return llvmAttrs;
+  }
+
+  LLVMMetadataList buildLLVMLoopAttributes(const uast::Loop* node) {
+    auto attrs = buildLLVMMetadataList(node->attributeGroup());
+    auto assertVectorize = buildAssertVectorize(node->attributeGroup());
+    if(assertVectorize) {
+      attrs.push_back(assertVectorize);
+    }
+    return attrs;
+  }
+
+
+
+
+  LLVMMetadataPtr buildAssertVectorize(const uast::AttributeGroup* node) {
```

Comment:
The need for this function made me pause a bit. Why do we need this function? Oh, is it because it is a syntactically different construct than `llvm.metadata`? On a very quick look the body of the loop in `buildLLVMMetadataList` looks pretty similar to this. Any chance of refactoring?